notion-core/src/main/scala/com/ambiata/notion/core/LineReducer.scala

```diff
@@ -0,0 +1,104 @@
+package com.ambiata.notion
+package core
+
+import java.security.MessageDigest
+
+import org.apache.commons.codec.binary.Base64
+
+import scala.collection.mutable.Queue
+import scalaz.{Foldable, Monoid}
+
+/**
+ * "Reducer-like" class to accumulate values while
+ * reading a file and transforming them to a final value
+ * when the whole file has been read
+ */
+trait LineReducer[T] {
+  type S
+  
+  /** initial value */
+  def init: S
+
+  /** reducer function using the current line and the current state S */
+  def reduce(line: String, s: S): S
+
+  /** final value from the last state */
+  def finalise(s: S): T
+}
+
+object LineReducer {
+
+  /** zip two line reducers together - aka parallel composition */
+  implicit class ZippedLineReducer[T1](r1: LineReducer[T1]) {
+    def zip[T2](r2: LineReducer[T2]): LineReducer[(T1, T2)] = new LineReducer[(T1, T2)] {
```

Comment:
I wanted to separate the essential definition from the useful builders/operators.